### ISSUE TITLE
Clingo Env: Conduit + Ascent + SENSEI w/ HDF5

### ISSUE THREAD

**Issue by ax3l** [LLNL/spack#32272]
### Steps to reproduce

Trying to concretize the following spec does lead to a Spack environment merge error:
```yaml
# spack.yaml
spack:
  specs:
  - adios2
  - ascent +adios2 +python
  - cmake
  - hdf5
  - lapackpp
  - mpi
  - python
  - py-cython
  - py-h5py
  - py-mpi4py
  - py-numpy
  - py-pip
  - py-setuptools
  - py-wheel
  - sensei +ascent ~catalyst +python

  packages:
    all:
      variants: +mpi ~fortran
    # default blocks at HDF5 1.8, resulting in unmergable solution
    conduit:
      variants: ~hdf5_compat
```
```console
spack env create vis-dev spack.yaml
spack env activate vis-dev
spack install
```

cc @cyrush @kwryankrattiger et al.

### Error message

```
==> Error: 552 fatal error(s) when merging prefixes:
    `/home/axel/src/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-10.3.0/hdf5-1.12.2-biyrbbrx4kgaupf7elfq7ti7oqtcqda7/.spack/install_environment.json` and `/home/axel/src/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-10.3.0/hdf5-1.12.2-fyvbywo56fyzjh5kinomhuz4cyt6gwio/.spack/install_environment.json` both project to `.spack/hdf5/install_environment.json`
    `/home/axel/src/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-10.3.0/hdf5-1.12.2-biyrbbrx4kgaupf7elfq7ti7oqtcqda7/.spack/install_manifest.json` and `/home/axel/src/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-10.3.0/hdf5-1.12.2-fyvbywo56fyzjh5kinomhuz4cyt6gwio/.spack/install_manifest.json` both project to `.spack/hdf5/install_manifest.json`
    `/home/axel/src/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-10.3.0/hdf5-1.12.2-biyrbbrx4kgaupf7elfq7ti7oqtcqda7/.spack/install_times.json` and `/home/axel/src/spack/opt/spack/linux-ubuntu20.04-skylake/gcc-10.3.0/hdf5-1.12.2-fyvbywo56fyzjh5kinomhuz4cyt6gwio/.spack/install_times.json` both project to `.spack/hdf5/install_times.json`
```

Somehow the conduit default-on `+hdf5_compat` forces part of the spec to be lowered to HDF5 1.8

```
$ spack spec | grep hdf5 | tr -d '  '
adios2@2.8.2%gcc@10.3.0+blosc+bzip2~cuda~dataman~dataspaces+fortran~hdf5~ipo+mpi+pic+png~python+shared+ssc+sst+sz+zfpbuild_type=Releasearch=linux-ubuntu20.04-skylake
^sz@2.1.12.2%gcc@10.3.0~fortran~hdf5~ipo~netcdf~pastri~python~random_access+shared~stats~time_compressionbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
^adios2@2.8.2%gcc@10.3.0+blosc+bzip2~cuda~dataman~dataspaces+fortran~hdf5~ipo+mpi+pic+png~python+shared+ssc+sst+sz+zfpbuild_type=Releasearch=linux-ubuntu20.04-skylake
^sz@2.1.12.2%gcc@10.3.0~fortran~hdf5~ipo~netcdf~pastri~python~random_access+shared~stats~time_compressionbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
^conduit@0.8.3%gcc@10.3.0~adios+blt_find_mpi~doc~doxygen~fortran+hdf5~hdf5_compat~ipo+mpi+parmetis+python+shared~silo+test~zfpbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
^hdf5@1.12.2%gcc@10.3.0~cxx~fortran~hl~ipo~java+mpi+shared~szip~threadsafe+toolsapi=defaultbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
hdf5
hdf5@1.12.2%gcc@10.3.0~cxx~fortran~hl~ipo~java+mpi+shared~szip~threadsafe+toolsapi=defaultbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
^hdf5@1.12.2%gcc@10.3.0~cxx~fortran+hl~ipo~java+mpi+shared~szip~threadsafe+toolsapi=defaultbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
sensei@4.0.0%gcc@10.3.0+adios2+ascent~catalyst+hdf5~ipo~libsim~miniapps+python+shared+vtkiobuild_type=RelWithDebInfopatches=9481e91,c4a671f,fdb490f,ff8957aarch=linux-ubuntu20.04-skylake
^adios2@2.8.2%gcc@10.3.0+blosc+bzip2~cuda~dataman~dataspaces+fortran~hdf5~ipo+mpi+pic+png~python+shared+ssc+sst+sz+zfpbuild_type=Releasearch=linux-ubuntu20.04-skylake
^sz@2.1.12.2%gcc@10.3.0~fortran~hdf5~ipo~netcdf~pastri~python~random_access+shared~stats~time_compressionbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
^conduit@0.8.3%gcc@10.3.0~adios+blt_find_mpi~doc~doxygen~fortran+hdf5+hdf5_compat~ipo+mpi+parmetis~python+shared~silo+test~zfpbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
^hdf5@1.8.22%gcc@10.3.0~cxx~fortran+hl~ipo+mpi+shared~szip~threadsafe+toolsapi=defaultbuild_type=RelWithDebInfoarch=linux-ubuntu20.04-skylake
```

### Information on your system

* **Spack:** 0.19.0.dev0 (90352d722358640e1b8d87f1125a0ee0675eb75e)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-skylake
* **Concretizer:** clingo

### Work-Around

If I add
```yaml
  concretizer:
    reuse: false
    unify: true
```

to my environment, then `clingo` finds a mergable solution.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

**Comment by adamjstewart** [comment]
I don't know anything about conduit, ascent, sensei, hdf5, or environments.

**Comment by ax3l** [comment]
@adamjstewart I added you because I think it's a concretizer issue - do you know who would be interested in that?

**Comment by adamjstewart** [comment]
@alalazo or @becker33

**Comment by ax3l** [comment]
Thanks a lot! :)

**Comment by kwryankrattiger** [comment]
I have found that putting things in the package preferences is mostly not useful in `unify:true` environments. That is probably a concretizer issue, but I think this will be mostly "fixed" with `package:require` going forward. #32260

To fix this now, I would recommend to use the `ecp-data-vis-sdk` package to handle the caveats of sensei/ascent/adios2/hdf5 for you. `ecp-data-vis-sdk +adios2 +ascent +hdf5 +sensei`. We have put a lot of effort in to make sure the member packages together as expected. Then downside is, those packages will be marked "implicit" by spack rather than explicit, but that will change soon-ish I hope. #31961

**Comment by alalazo** [comment]
The issue in my opinion are the default of environments, which might lead to these conflicts. We have PRs to fix those in v0.19. I don't see by just reading the description a clear problem with the concretizer. If specs are concretized separately and with reuse what you see here might happen.

A couple of notes:
- The error seems to be due to having two versions of `hdf5@1.12.2` with different hashes, and `hdf5@1.8.22` is a different issue. 
- If `reuse` is true concretization is stateful, and depends on whatever is already installed locally or available from a registered binary mirror - but that is the intended behavior.

**Comment by ax3l** [comment]
Thank you both, I think that solves it for now.

I'll roll with `ecp-data-vis-sdk +adios2 +ascent +hdf5 +sensei` as recommended :)
